presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/BingTile.java

```diff
@@ -145,20 +157,83 @@ public String toQuadKey()
+    }
+
+    public List<BingTile> findChildren(int newZoom)
+    {
+        if (newZoom == zoomLevel) {
+            return ImmutableList.of(this);
+        }
+
+        checkArgument(newZoom <= MAX_ZOOM_LEVEL, "newZoom must be less than or equal to %s: %s", MAX_ZOOM_LEVEL, newZoom);
+        checkArgument(newZoom >= zoomLevel, "newZoom must be greater than or equal to current zoom %s: %s", zoomLevel, newZoom);
+
+        int zoomDelta = newZoom - zoomLevel;
+        int xNew = x << zoomDelta;
+        int yNew = y << zoomDelta;
+        ImmutableList.Builder<BingTile> builder = ImmutableList.builderWithExpectedSize(1 << (2 * zoomDelta));
+        for (int yDelta = 0; yDelta < 1 << zoomDelta; ++yDelta) {
+            for (int xDelta = 0; xDelta < 1 << zoomDelta; ++xDelta) {
+                builder.add(BingTile.fromCoordinates(xNew + xDelta, yNew + yDelta, newZoom));
+            }
+        }
+        return builder.build();
+    }
+
+    public BingTile findParent()
+    {
+        return findParent(zoomLevel - 1);
+    }
+
+    public BingTile findParent(int newZoom)
+    {
+        if (newZoom == zoomLevel) {
+            return this;
+        }
+
+        checkArgument(newZoom >= 0, "newZoom must be greater than or equal to 0: %s", newZoom);
+        checkArgument(newZoom <= zoomLevel, "newZoom must be less than or equal to current zoom %s: %s", zoomLevel, newZoom);
+
+        int zoomDelta = zoomLevel - newZoom;
+        return BingTile.fromCoordinates(x >> zoomDelta, y >> zoomDelta, newZoom);
+    }
+
     /**
-     * Encodes Bing tile as a 64-bit long: 23 bits for X, followed by 23 bits for Y,
-     * followed by 5 bits for zoomLevel
+     * Encodes Bing tile as a 64-bit long:
+     * Version (5 bits), 0 (4 bits), x (23 bits), Zoom (5 bits), 0 (4 bits), y (23 bits)
+     * (high bits left, low bits right).
+     * <p>
+     * This arrangement maximizes low-bit entropy for the Java long hash function.
      */
     public long encode()
     {
-        return (((long) x) << 28) + (y << 5) + zoomLevel;
+        // Java's long hash function just XORs itself right shifted 32.
+        // This is used for bucketing, so if you have 2^k buckets, this only
+        // keeps the k lowest bits.  This puts the highest entropy bits
+        // (finest resolution x and y bits) in places that contribute to the
+        // low bits of the hash.
+        return (((long) VERSION << VERSION_OFFSET) | y | ((long) x << 32) | ((long) zoomLevel << ZOOM_OFFSET));
     }
 
     public static BingTile decode(long tile)
     {
-        int tileX = (int) (tile >> 28);
-        int tileY = (int) ((tile % (1 << 28)) >> 5);
-        int zoomLevel = (int) (tile % (1 << 5));
+        int version = (int) (tile >>> VERSION_OFFSET) & BITS_5;
+        if (version == 0) {
+            return decodeV0(tile);
+        }
+        else {
+            throw new IllegalArgumentException(format("Unknown Bing Tile encoding version: %s", version));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: version (declared at line 223). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 228. Consider formatting this data with &#39;%d&#39; instead.